### ISSUE TITLE
Kawase blur

### ISSUE THREAD

**Issue by osleg** [yshui/picom#32]
Is it possible to have kawasi blend mode with your fork?
https://github.com/tryone144/compton

**Comment by tryone144** [comment]
My latest code (the [improved_fbo](https://github.com/tryone144/compton/tree/improved_fbo) branch) is based on commit 316eac0613bf342ff91cc645a6c3c80e6b9083fb and already includes some rendering changes.

I will look into your `next` branch when I find some time. It looks like you moved some opengl related code around, so rebasing might take a while. Once this is done I can post a pull-request.

**Comment by yshui** [comment]
@tryone144 So I am in the process of moving all backend related code into separate modules, but that is coming along slowly. Apparently that is going to cause more conflicts. So for users' best interesting it is better to merge your code in first. But that will make my job harder.

What do you think is the better way to do this?

**Comment by tryone144** [comment]
@yshui Currently my fork contains both blur methods, which will result in at least twice the amount of work needed when you start refactoring more opengl stuff. I basically duplicated most of the shader generation and rendering code.
Furthermore, I changed both blur methods to use one or more global screensize fbos to improve performance in the *dual kawase* mode. This meant some changes to the glx render cache structures.

If you plan on putting the blur related code into a seperate module, I'd suggest merging the fbo improvements first and then add the new blur method.
Removing the old blur shader might help cleanup but this would be a breaking change. So not optimal for the user.

I'd like to first take a look at the current changes and see if I can merge them rather easily. Hopefully I can get to it this week.

**Comment by osleg** [comment]
If I might suggest... 
AFAIK the old blur method is not used by majority of people since it kind bad -ish.
So removing the old blurring method sounds like a good idea to me.

Just my few cents.
Unfortunately I'm supper rusty in C and don't have much free time to help :(

**Comment by frebib** [comment]
I've been using the improved blur for a couple of weeks now and it not only looks better, it performs a lot faster than the equivalent blur level on intel integrated.

Just to keep track of this, @tryone144 you don't have issues enabled so I'm mentioning this here:
The `blur-background-fixed = false` option seems to be broken with this update which makes blurred shadows jump in and out instead of fade as they used to. I don't know how difficult of a fix  it would be but it'd be nice to see that fixed before it is merged here.

Thanks to both of you for your work on compton, I'm glad it's getting more love :+1:

**Comment by tryone144** [comment]
@frebib Changing the "blur strength" on the fly is currently not implemented. The old code actually hacked this effect by giving the center pixel more weight. The new algorithm works differently and would require some major changes in the render passes to implement a strength-fade effect.

**Comment by yshui** [comment]
@tryone144 we could simply paint the blurred picture with alpha, right?

That's what the new xrender backend will do.

**Comment by tryone144** [comment]
This may work. I'll take a look into it.Is the new xrender code already up somewhere?On 15 Nov 2018 15:52, yshui <notifications@github.com> wrote:@tryone144 we could simply paint the blurred picture with alpha, right?
That's what the new xrender backend will do.

**Comment by yshui** [comment]
@tryone144 it's in the `split-backends` branch.

**Comment by frebib** [comment]
It seems to be doing the opposite, strangely. I can't quite tell if the background is still blurred but I can clearly read text through transparent windows. Thanks for looking into it!

Before:
![](https://frebib.net/s/screenshot-2018-11-15-193424.png)

After:
![](https://frebib.net/s/screenshot-2018-11-15-193554.png)

Windows also disappear almost immediately whereas before they faded out slowly.

Perhaps you'd like to try my config and see for yourself? https://raw.githubusercontent.com/frebib/dotfiles/master/.config/compton.conf

**Comment by tryone144** [comment]
Hmm, interesting...

Fading speed should not be changed at all, only opacity of the blur itself will change at the same rate as window opacity. The difference is, that a window with opacity 0 is not blurred at all and one with 1 is blurred as before. At a lower `blur-strength` level (e.g. 2) the result looks like the current `xrender` backend.

Your screenshots look exactly like expected: The bottom one is less blurred.
It is not really clear to me, what you are trying to achieve. Feel free to open a new issue over at https://github.com/tryone144/compton as to not bloat this issue more. ;)

**Comment by alex47** [comment]
> @frebib Changing the "blur strength" on the fly is currently not implemented. The old code actually hacked this effect by giving the center pixel more weight. The new algorithm works differently and would require some major changes in the render passes to implement a strength-fade effect.

strength-fade effect is sadly not possible with this blur (or it would require extreme level of hacking)
it would work for a certain amount but then it would suddenly jump in quality, kinda like changing gears while driving a car

**Comment by yshui** [comment]
Sorry but I will have to mark these discussions off-topic @alex47 @frebib

**Comment by c0nd3v** [comment]
Please can we get this implemented.

**Comment by tryone144** [comment]
How's the progress of the new backends, @yshui? Which branch is best to build my changes on?
If I see correctly, the new backends are already present in `next` but only available via the `--experimental-backends` switch.

**Comment by yshui** [comment]
@tryone144 The new backends are mostly working, except a few crashes. The backend API is pretty much settled now. If you want to start your work, you can base it off `next`.

Thanks!

> If I see correctly, the new backends are already present in next but only available via the --experimental-backends switch.

You are correct.

**Comment by ronasimi** [comment]
Have you considered making Kawase the default blur method once this is merged?  It seems to be more CPU/GPU friendly than the old default blur method.

**Comment by prankousky** [comment]
+1 for kawase blur. I tried the tyrone fork, but don't like it. I'd rather stay with this version but have intense blur :)

**Comment by alex47** [comment]
@prankousky what's the difference between the tryone fork and this intense blur?
how is this better?

**Comment by edward-shen** [comment]
The kawase blur is an significantly less computationally intensive effect while retaining the same visual effects.

**Comment by Anarky** [comment]
@tryone144 If you are still planning to rebase your work, would it be possible to add an option to use the wallpaper/desktop as the blur instead of what's behind the application?
While it looks great on tiled applications, I usually use 3/4 windows in fullscreen and it doesn't look as pretty.
Thanks!

**Comment by tryone144** [comment]
> would it be possible to add an option to use the wallpaper/desktop as the blur instead of what's behind the application?

This seems to be a good feature, but is not necessarily linked to the used blur-algorithm.
As such opening a new issue might be better. :wink:

**Comment by anesuc** [comment]
Hey, whats the progress on this? I am working on a project that heavily relies on this! Currently using the other fork, but would be good to use this version instead as the other one is behind by a bit

Thanks!

**Comment by Rapiz1** [comment]
Any progress?

**Comment by frebib** [comment]
https://github.com/tryone144/compton/tree/feature/dual_kawase

**Comment by tryone144** [comment]
> https://github.com/tryone144/compton/tree/feature/dual_kawase

This branch is based on the current [next](https://github.com/yshui/picom/tree/next) branch and should implement the new *dual_kawase* blur algorithm with experimental-backends.

Beware that this is WIP and still needs a bit of cleanup. I am waiting on https://github.com/yshui/picom/pull/286 for the cli interface. Once this is complete and local testing doesn't reveal any significant problem, I'll open a PR.

In the meantime, if you aren't worried about your cat (or other pet) you can test it at your own risk and report any issues over at https://github.com/tryone144/compton. ;)

**Comment by tbkr** [comment]
Hey, this probably isn't the right place to ask this but is there any possibility to darken the kawase_dual blur in a way that a transparent background of e.g. a terminal is a bit darker?

I have a lot of bright background images and it would be nice if there is an option which allows to additionally dim the blur. I'm currently not really happy with the contrast ratio. I know that there is an option for dimming inactive windows but as far as I know it's not quite what I want to achieve because it also dims the window content.

I'm aware that setting a custom kernel for blurring could achieve the same effect if all values don't add up to 1. 

Has anybody stumbled across the same problem?

**Comment by SitiSchu** [comment]
@tbkr its offtopic but you can use a shadow to achieve that effect (for example with `shadow-opacity = 0.4;`)
![image](https://user-images.githubusercontent.com/18517178/75067991-90696f80-54ee-11ea-8c5a-fe0d103a2532.png)

**Comment by tbkr** [comment]
@SitiSchu isn't the shadow only applied outside of a window? In my case the tint of a blurred background isn't really affected by the shadow. Also changing the alpha value only results in a darker or brighter shadow outside the window.

Even setting the opacity to 1 results in no additional darkening of the blurred window background.

Did I get you wrong?

**Comment by manilarome** [comment]
> isn't the shadow only applied outside of a window? 

It can, check `full-shadow` in the manual.

**Comment by tbkr** [comment]
@manilarome Thanks, that indeed did the trick.

**Comment by thezealousfool** [comment]
@tryone144  Updates on Kawase blur?

**Comment by Aire-One** [comment]
Sorry for the "stack overflow style" question, but I'm genuinely wondering:

Wouldn't glsl shaders be a better way to implement dual kawase? Or is there some limitation I'm not aware to?

Regarding firsts post in #295, it seems to fit the use case to me.

**Comment by tryone144** [comment]
@Aire-One The dual-kawase algorithm needs more than just a single glsl shader.

The blur itself consist of two glsl shaders that are called for multiple times in succession to "downscale" and "upscale" the blurred image. This is not possible with the glx-fshader-win option.

**Comment by Aire-One** [comment]
Thanks for the quick answer @tryone144 :)

**Comment by mighty9245** [comment]
@tryone144, so what's up with implementing dual kawase blurring algorithm? this night i tried to port your first implementation, but when things were mostly done i found this issue and that you already have done this :)

@yshui, a question to you that has been asked but not answered before: once we have dual kawase blurring algorithm, can we make things simpler and get rid of other blurring algorithms? to be honest i don't see a reason to use other blurring algorithms instead of dual kawase

**Comment by yshui** [comment]
@mighty9245 
> once we have dual kawase blurring algorithm, can we make things simpler and get rid of other blurring algorithms?

_maybe_. I don't really think removing other blur algorithm will make things that simpler.

**Comment by mighty9245** [comment]
let me explain my point

* as a user:

dual kawase should be more efficient and cpu-/gpu-friendly than other blurring algorithms. i don't have any proofs and haven't done any benchmarks, but alex47, who integrated this blurring algorithm into kde (and as far as i know it replaced the existing one), claims that. 

dual kawase overall produces better looking blur than other blurring algorithms. nevertheless lowering it's stength makes it to mimic other blurring algorithms.

another thing is configuration simplicity. dual kawase is configured with literally a single option - strength (though there are two option hidden behind it - number of iterations and offset).

given these three points i don't see a reason to use other blurring algorithms instead of dual kawase. 

* as a dev:

there is not much good things in having a zoo of blurring algorithms. 

currently we have two backends: xrender and glx, and we may have more backends in future (e. g. egl, vulkan). we must support all the backends and all the backends must have same features (if it's possible with a particular backend, of course).

we need to implement all these blurring algorithms first, so we basically need to implement the same n things for m times, where n number of blurring algorithms and m number of backends.

if we are doing something with blur, we probably need to reflect our change to all blurring algorithms for all backends. implement the same n things for m times.

and every blurring algorithms may have it's own caveats. every backend may have it's own caveats. every combination of blurring algorithm and backend may have it's own caveats.

btw, current blur implementation looks quite messy for me. because of current blurring algorithms are similiar, they are sharing the same routine with if's and some small routines for a particular blurring algorithm. maybe it's efficient, but it makes harder to understand what is going on until you spend some time looking at code and tracing where everything goes. looking through the code makes you think "how these blurring algorithms are different? they're sharing the same routine and doing the same things!".

**Comment by tryone144** [comment]
@mighty9245 
Dual-filter kawase blur is significantly easier on computations at high blur-strengths compared to a plain gauss-kernel blur.

With my patch, there are effectively **two** different blur algorithms. And this only holds true for the new experimental gl/glx backend. The xrender backend only has support for blur-kernels. Adding support there would require a fast CPU-based implementation of the kawase filter.
So we can't really drop the kernel-based blur-options without dropping blur in xrender.

Other than that, the *gaussian* and *box* blur methods are just config-sugar which automatically generate a suitable kernel for the kernel-blur algorithm. They don't really add any complexity to the main blur-code, whereas dual-filter kawase requires its own init and render routines.

**Comment by Diegovsky** [comment]
Is it going to be implemented? I'm curious to see how this old boi will handle the lighter kawase blur

**Comment by Shadorain** [comment]
> @tbkr its offtopic but you can use a shadow to achieve that effect (for example with `shadow-opacity = 0.4;`)
> ![image](https://user-images.githubusercontent.com/18517178/75067991-90696f80-54ee-11ea-8c5a-fe0d103a2532.png)

Slightly off topic, but dude I can't handle how nice your wallpaper and terminal theme is lol with the blur, it looks so dang good! Would you be able to send me the wallpaper :sweat_smile: I'd appreciate it alot lol, sorry to bother. 
discord: Shadorain#4182

**Comment by AryToNeX** [comment]
Will this blur be available to applications who request it via a X window property (much like KWin does?)

**Comment by mighty9245** [comment]
@AryToNeX, in theory, you should be able to configure it in this way using `blur-background-exclude`.

**Comment by AryToNeX** [comment]
@mighty9245 Do you have an example of sorts? I have been requested to support picom's blur in a project of mine

**Comment by mighty9245** [comment]
@AryToNeX, oh, that changes the situation. i thought you want to configure your environment so picom will blur background only of windows with a special property. well, picom should blur all windows automatically if they have a bit of transparency and not excluded with configured rules. so basically what you app should do is to be a bit transparent and if blurring is enabled picom will automatically deal with it unless your app is somehow excluded. it's almost completely on the users side.

**Comment by nick87720z** [comment]
Sorry if writing not at right issue. Before reading what kawase filter is, I remembered about some fast blur, probably from xcompmgr - based in multiple box filters. It so happened, that I reinvented moving averages way for box filter before reading any articles. By mistake I assumed, that so much mentioned kawase filter is that old fast filter).

To the point, I was about to propose following:
- existing box experimental blur could use moving averages and
- add option for passes number. Of course could be that it will be used with nothing but box, so could be made specific box option to better optimize it.

I noticed (learning code to find, how blur is implemented), that xrender also has blur (only kernel). Imho, this is worst possible way from performance aspect, so would be very useful if this backend gets box filter using moving avg. With multiple passes it could so happen, that both gauss and kernel filter will be rarely used (unless one wants interesting effects, requiring custom composed kernels, e.g. chess-like, circle/square stroke (all values at kernel edge with zeros in the middle).

**Comment by mighty9245** [comment]
@nick87720z, xrender backend is cpu-based. we don't have ability to use shaders there.

**Comment by nick87720z** [comment]
I mentioned glsl with second url, because they could be useful for glx, if not yet (I'm not sure, wether it requires opencl support).

As for xrender - hm, I did not analyze yet existing kernel method. But anyway, if it's cpu based, than moreover it needs improvement (lol, I'm writing about it without even trying in action).

**Comment by AntouanK** [comment]
just curious, is there any update on this?
using tryone's fork right now, that blur looks nice :D
wondering if this fork will get it as well.

**Comment by tryone144** [comment]
See #382.
I'm currently a bit short on time, but should be ready for merging soon™. :wink:

**Comment by Qwerty-Space** [comment]
> 
> 
> I'm currently a bit short on time, but should be ready for merging soon™. 😉

Hopefully within the next few weeks!  I'm making the jump to Linux on my main PC once all of the parts arrive :-)

**Comment by Joan31** [comment]
Thank you a lot for your work.
I'm waiting for this new feature and have a nice blur effect with kawase :)

**Comment by Qwerty-Space** [comment]
Any chance of this getting merged soon?

**Comment by tryone144** [comment]
#382 has been merged into `next` :tada:
You can now use dual-filter kawase blur by setting `--blur-method dual_kawase` and changing strength with `--blur-strength LEVEL` where LEVEL is an integer between 1 and 20 (inclusive).

For now, this is only implemented for the [new experimental backends](https://github.com/yshui/picom#--experimental-backends) you can use with `--experimental-backends`.
If you can't use the experimental opengl backend and absolutely have to use the legacy backend, please leave a message. Please understand that since there are quite a few fundamental changes involved, I'll only backport the current implementation to the legacy backend if there really is a strong need for it.
